### PR TITLE
Round-trip OpenType tags through BCP 47

### LINKED ISSUE #2380: Round-tripping OpenType tags through BCP 47 losslessly
Firefox [calls `hb_ot_tag_to_language` and `hb_buffer_set_language`](https://hg.mozilla.org/mozilla-central/file/2b0e2483e2eaaef7ac49866a98e8e857520bdd5c/gfx/thebes/gfxHarfBuzzShaper.cpp#l1393) to implement [`font-language-override`](https://drafts.csswg.org/css-fonts-4/#propdef-font-language-override), which causes [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1346261): the tag is parsed case-insensitively but CSS says the string is case-sensitive.

I would like to fix this by adding a new private use subtag syntax to HarfBuzz: x-hbot-<var>AABBCCDD</var>, where each pair of letters in the trailing subtag represents a byte of the tag. For example, `hb_ot_tag_to_language` would convert the tag 'trk ' (which is _not_ the right tag for Turkish) to x-hbot-74726b20, which `hb_ot_tags_from_script_and_language` would later interpret as 'trk ' instead of [normalizing the case to 'TRK '](https://github.com/harfbuzz/harfbuzz/blob/e53c44e3260d6136e27635b5bbcb2da43bc4f1b1/src/hb-ot-tag.cc#L322-L323).

@behdad @jfkthame What do you think?

### PR DESCRIPTION
**Pull request by dscorbett**
Closes #2380.

### CODE REVIEW

File: src/hb-ot-tag.cc [hunk 1]
```diff
@@ -495,10 +514,10 @@ hb_ot_tags_to_script_and_language (hb_tag_t       script_tag,
     *language = hb_ot_tag_to_language (language_tag);
     if (script_count == 0 || primary_script_tag[0] != script_tag)
     {
-      unsigned char *buf;
+      char *buf;
       const char *lang_str = hb_language_to_string (*language);
       size_t len = strlen (lang_str);
-      buf = (unsigned char *) malloc (len + 11);
+      buf = (char *) malloc (len + 16);
```
Comment (ebraminio) [inline src/hb-ot-tag.cc:520]:
```
=================================================================
==22304==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x603000004972 at pc 0x00000043b114 bp 0x7fffd5454240 sp 0x7fffd54539d8
WRITE of size 15 at 0x603000004972 thread T0
    #0 0x43b113 in vsprintf ??:?
    #1 0x43c093 in __interceptor_sprintf ??:?
    #2 0x7fe469fbbf21 in hb_ot_tags_to_script_and_language /root/project/src/hb-ot-tag.cc:532:2
    #3 0x4c8d3c in test_tags_to_script_and_language /root/project/test/api/test-ot-tag.c:261:3
    #4 0x4c752c in test_ot_tags_to_script_and_language /root/project/test/api/test-ot-tag.c:269:3
    #2 0x7fe469ce5c3d in g_test_summary ??:?
    #3 0x7fe469ce59e3 in g_test_summary ??:?
    #4 0x7fe469ce59e3 in g_test_summary ??:?
    #5 0x7fe469ce59e3 in g_test_summary ??:?
    #6 0x7fe469ce59e3 in g_test_summary ??:?
    #7 0x7fe469ce59e3 in g_test_summary ??:?
    #8 0x7fe469ce6129 in g_test_run_suite ??:?
    #9 0x7fe469ce6144 in g_test_run ??:?
    #13 0x4c8008 in hb_test_run /root/project/test/api/./hb-test.h:88:10
    #14 0x4c6e6d in main /root/project/test/api/test-ot-tag.c:541:10
    #10 0x7fe4697151e2 in __libc_start_main ??:?
    #11 0x41c3ed in _start ??:?

0x603000004972 is located 0 bytes to the right of 18-byte region [0x603000004960,0x603000004972)
allocated by thread T0 here:
    #0 0x494b4d in __interceptor_malloc ??:?
    #1 0x7fe469fbbe50 in hb_ot_tags_to_script_and_language /root/project/src/hb-ot-tag.cc:520:22
    #2 0x4c8d3c in test_tags_to_script_and_language /root/project/test/api/test-ot-tag.c:261:3
    #3 0x4c752c in test_ot_tags_to_script_and_language /root/project/test/api/test-ot-tag.c:269:3
    #1 0x7fe469ce5c3d in g_test_summary ??:?

SUMMARY: AddressSanitizer: heap-buffer-overflow (/root/project/test/api/.libs/test-ot-tag+0x43b113) in vsprintf
Shadow bytes around the buggy address:
  0x0c067fff88d0: fa fa fd fd fd fa fa fa fd fd fd fa fa fa 00 00
  0x0c067fff88e0: 00 00 fa fa fd fd fd fd fa fa fd fd fd fd fa fa
  0x0c067fff88f0: fd fd fd fa fa fa fd fd fd fa fa fa 00 00 00 00
  0x0c067fff8900: fa fa fd fd fd fa fa fa fd fd fd fd fa fa fd fd
  0x0c067fff8910: fd fa fa fa fd fd fd fa fa fa 00 00 00 00 fa fa
=>0x0c067fff8920: 00 00 00 fa fa fa fd fd fd fa fa fa 00 00[02]fa
  0x0c067fff8930: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8940: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8950: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8960: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c067fff8970: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==22304==ABORTING
FAIL test-ot-tag (exit status: 1)
```

Guess you need to just make more space for it? Thanks

**Comment by behdad** [comment]
Thanks.  While the code is correct, it looks very crude and 80s.  For example, the hex digit read / write can eeeeasily be moved to inline functions...

But hey, why not use sprintf for writing the hex out, and possibly even using sscanf to read?  sscanf has superpower capabilities...

**Comment by jfkthame** [comment]
> But hey, why not use sprintf for writing the hex out, and possibly even using sscanf to read? sscanf has superpower capabilities...

I suspect linking with sscanf may pull in quite a lot of code (as it has to include the code for every possible conversion spec...), which seems a bit regrettable just to read some hex digits in a niche code-path.

**Comment by dscorbett** [comment]
I did try to use `sprintf` in `hb_ot_tags_to_script_and_language` originally, as you can see from the “force-pushed” link above. Various sanitizer checks failed, so I reimplemented it this more explicit way.

I intentionally avoided `sscanf` because most conversions skip white space, which would not be appropriate here.

**Comment by behdad** [comment]
Thanks.  Also, no `sprintf`.  Always `snprintf` only.

**Comment by behdad** [comment]
So yeah, generous use of inline functions at least please.

**Comment by dscorbett** [comment]
I added some inline functions and some more tests. How is it now?

File: src/hb-ot-tag.cc [hunk 2]
```diff
@@ -434,30 +448,28 @@ hb_ot_tag_to_language (hb_tag_t tag)
...
-      }
+      buf[0] = TOLOWER (tag >> 24);
+      buf[1] = TOLOWER ((tag >> 16) & 0xFF);
+      buf[2] = TOLOWER ((tag >> 8) & 0xFF);
+      buf[3] = '-';
+      str += 4;
     }
-    buf[10] = '\0';
-    return hb_language_from_string (str, -1);
+    sprintf (str, "x-hbot-%02x%02x%02x%02x", HB_UNTAG (tag));
```
Comment (ebraminio) [inline src/hb-ot-tag.cc:471]: `snprintf` was suggested for here IIRC

File: src/hb-algs.hh [hunk 3]
```diff
@@ -577,6 +577,12 @@ static inline unsigned char TOUPPER (unsigned char c)
 { return (c >= 'a' && c <= 'z') ? c - 'a' + 'A' : c; }
 static inline unsigned char TOLOWER (unsigned char c)
 { return (c >= 'A' && c <= 'Z') ? c - 'A' + 'a' : c; }
+static inline bool ISHEX (unsigned char c)
+{ return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F'); }
+static inline unsigned char FROMHEX (uint8_t c)
+{ return (c & 0xF) <= 9 ? (c & 0xF) + '0' : (c & 0xF) + 'a' - 10; }
+static inline uint8_t TOHEX (unsigned char c)
```
Comment (behdad) [inline src/hb-algs.hh:584]: The names of the above sounds wrong?  FROMHEX should take char and return number.

Comment (behdad) [reply src/hb-algs.hh:584]: Am not sure about the casting but am fine keeping.

File: src/hb-ot-tag.cc [hunk 4]
```diff
@@ -434,30 +448,28 @@ hb_ot_tag_to_language (hb_tag_t tag)
...
-      }
+      buf[0] = TOLOWER (tag >> 24);
+      buf[1] = TOLOWER ((tag >> 16) & 0xFF);
+      buf[2] = TOLOWER ((tag >> 8) & 0xFF);
+      buf[3] = '-';
+      str += 4;
     }
-    buf[10] = '\0';
-    return hb_language_from_string (str, -1);
+    snprintf (str, 16, "x-hbot-%02x%02x%02x%02x", HB_UNTAG (tag));
```
Comment (behdad) [inline src/hb-ot-tag.cc:471]: Why not just `%08x`?

**Comment by behdad** [comment]
We generate the hex _only_ if the casing will be lost, right?

**Comment by dscorbett** [comment]
No: the current format is still parsed but only the new format is generated.

**Comment by behdad** [comment]
> No: the current format is still parsed but only the new format is generated.

Got it.  Thanks.

File: src/hb-ot-tag.cc [hunk 5]
```diff
@@ -319,12 +319,26 @@ parse_private_use_subtag (const char     *private_use_subtag,
...
+  if (s[0] == '-') {
+    s += 1;
+    char c;
+    for (i = 0; i < 8 && ISHEX (s[i]); i++)
+    {
+      c = FROMHEX (s[i]);
+      if (i % 2 == 0)
+	tag[i / 2] = c << 4;
+      else
+	tag[i / 2] += c;
```
Comment (behdad) [inline src/hb-ot-tag.cc:331]: The classic way would have been to do `tag = (tag << 4) | FROMHEX (s[i])`.

Comment (behdad) [reply src/hb-ot-tag.cc:331]: Oh I see the problem.  Nvmd.  Good enough for me.

**Pull request merged**